sw/device/lib/dif/dif_uart.c

```
@@ -0,0 +1,409 @@
+  size_t bytes_written = 0;
+  while ((bytes_written < bytes_requested) && !uart_tx_full(uart)) {
+    uart_write_tx_fifo(uart, data[bytes_written]);
+    ++bytes_written;
+  }
+
+  return bytes_written;
+}
+
+// Read up to |bytes_requested| number of bytes from the RX FIFO
+static size_t uart_receive_bytes(const dif_uart_t *uart, size_t bytes_requested,
+                                 uint8_t *data) {
+  size_t bytes_read = 0;
+  while ((bytes_read < bytes_requested) && !uart_rx_empty(uart)) {
+    data[bytes_read] = uart_read_rx_fifo(uart);
+    ++bytes_read;
+  }
+
+  return bytes_read;
+}
+
+bool dif_uart_init(uintptr_t base_addr, const dif_uart_config_t *config,
+                   dif_uart_t *uart) {
+  if (uart == NULL || config == NULL || base_addr == 0) {
+    return false;
+  }
+
+  uart->base_addr.inner_ptr = (uint32_t *)base_addr;
+
+  return uart_configure(uart, config);
+}
+
+bool dif_uart_configure(const dif_uart_t *uart,
+                        const dif_uart_config_t *config) {
+  if ((uart == NULL) || (config == NULL)) {
+    return false;
+  }
+
+  return uart_configure(uart, config);
+}
+
+bool dif_uart_set_watermark_rx(const dif_uart_t *uart,
+                               dif_uart_watermark_t watermark) {
+  if (uart == NULL) {
+    return false;
+  }
+
+  // Check if the requested watermark is valid, and get a corresponding
+  // register definition to be written
+  uint32_t rxi_level;
+  switch (watermark) {
+    case kDifUartWatermarkByte1:
+      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL1;
+      break;
+    case kDifUartWatermarkByte4:
+      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL4;
+      break;
+    case kDifUartWatermarkByte8:
+      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL8;
+      break;
+    case kDifUartWatermarkByte16:
+      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL16;
+      break;
+    case kDifUartWatermarkByte30:
+      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL30;
+      break;
+    default:
```

You don't need a `default` if all the variants are covered, and in fact it may hide problems in future (if more variants are added).

---

Done, the only annoying bit now is that because warnings==errors, I have to initialise rxi_level before the switch.

---

@lenary FWIW I don't think having/not having `default` matters for future-proofing. C treats enums as a dumb alias over an integral type, unlike in C++, where you can do exhaustive switching.

---

To be honest I don't like the fact that now I need to do something like this:
`uint32_t rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL1;`

before, to address the Werror that occurs because compiler thinks that `rxi_level` may be uninitialised.
```
  switch (watermark) {
    case kDifUartWatermarkByte1:
      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL1;
      break;
    case kDifUartWatermarkByte4:
      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL4;
      break;
    case kDifUartWatermarkByte8:
      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL8;
      break;
    case kDifUartWatermarkByte16:
      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL16;
      break;
    case kDifUartWatermarkByte30:
      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL30;
      break;
  }
```
I think this would be another argument for returning error code enums, and would be better than current solution of removing the default.

@mcy @lenary your thoughts?

---

For the moment, put `default` back in. We can always revisit this decision later (not that I think we will).